App boot fails with LDAP enabled

<!--
Thanks for reporting a problem with Airsonic! Please complete all of the
following sections to make it easier to assist you.
-->

## Problem description

I decided to give this a go and started up the latest docker (airsonicadvanced/airsonic-advanced:latest@sha256:6af83b at time of writing) with a clean config directory (not using my existing Airsonic database). I use LDAP to authenticate with Airsonic normally so I logged and set that up. Upon restarting the container to pick up the LDAP changes it fails to boot with the following error:
```
Failed to instantiate [org.airsonic.player.security.JWTRequestParameterProcessingFilter]: Factory method 'jwtAuthFilter' threw exception; nested exception is java.lang.IllegalStateException: Cannot apply org.springframework.security.config.annotation.authentication.configurers.ldap.LdapAuthenticationProviderConfigurer@773c2214 to already built object
```

### Steps to reproduce

1. *Run docker container*
2. *Enable LDAP in web console*
3. *Stop and start the docker container to pick up LDAP settings*
4. *Airsonic no longer starts*

## System information

 * **Airsonic version**: 6af83b
 * **Operating system**: Docker on Ubuntu 20.04
 * **Java version**: Docker packaged
 * **Proxy server**: Traefik 2.0
 * **Client**: Firefox 75.0
 * **Language**: US English

## Additional notes

Full stack trace:

```
2020-05-07 15:12:12.587 ERROR --- o.s.boot.SpringApplication               : Application run failed,
,
org.springframework.context.ApplicationContextException: Unable to start web server; nested exception is org.springframework.boot.web.server.WebServerException: Unable to start embedded Tomcat,
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.onRefresh(ServletWebServerApplicationContext.java:156) ~[spring-boot-2.2.6.RELEASE.jar!/:2.2.6.RELEASE],
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:544) ~[spring-context-5.2.5.RELEASE.jar!/:5.2.5.RELEASE],
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:141) ~[spring-boot-2.2.6.RELEASE.jar!/:2.2.6.RELEASE],
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:747) ~[spring-boot-2.2.6.RELEASE.jar!/:2.2.6.RELEASE],
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:397) ~[spring-boot-2.2.6.RELEASE.jar!/:2.2.6.RELEASE],
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:315) ~[spring-boot-2.2.6.RELEASE.jar!/:2.2.6.RELEASE],
	at org.springframework.boot.builder.SpringApplicationBuilder.run(SpringApplicationBuilder.java:140) ~[spring-boot-2.2.6.RELEASE.jar!/:2.2.6.RELEASE],
	at org.airsonic.player.Application.main(Application.java:70) ~[classes!/:11.0.0-SNAPSHOT],
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na],
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source) ~[na:na],
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source) ~[na:na],
	at java.base/java.lang.reflect.Method.invoke(Unknown Source) ~[na:na],
	at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:48) ~[airsonic.war:11.0.0-SNAPSHOT],
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:87) ~[airsonic.war:11.0.0-SNAPSHOT],
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:51) ~[airsonic.war:11.0.0-SNAPSHOT],
	at org.springframework.boot.loader.WarLauncher.main(WarLauncher.java:58) ~[airsonic.war:11.0.0-SNAPSHOT],
Caused by: org.springframework.boot.web.server.WebServerException: Unable to start embedded Tomcat,
	at org.springframework.boot.web.embedded.tomcat.TomcatWebServer.initialize(TomcatWebServer.java:126) ~[spring-boot-2.2.6.RELEASE.jar!/:2.2.6.RELEASE],
	at org.springframework.boot.web.embedded.tomcat.TomcatWebServer.<init>(TomcatWebServer.java:88) ~[spring-boot-2.2.6.RELEASE.jar!/:2.2.6.RELEASE],
	at org.springframework.boot.web.embedded.tomcat.TomcatServletWebServerFactory.getTomcatWebServer(TomcatServletWebServerFactory.java:438) ~[spring-boot-2.2.6.RELEASE.jar!/:2.2.6.RELEASE],
	at org.springframework.boot.web.embedded.tomcat.TomcatServletWebServerFactory.getWebServer(TomcatServletWebServerFactory.java:191) ~[spring-boot-2.2.6.RELEASE.jar!/:2.2.6.RELEASE],
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.createWebServer(ServletWebServerApplicationContext.java:180) ~[spring-boot-2.2.6.RELEASE.jar!/:2.2.6.RELEASE],
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.onRefresh(ServletWebServerApplicationContext.java:153) ~[spring-boot-2.2.6.RELEASE.jar!/:2.2.6.RELEASE],
	... 15 common frames omitted,
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'jwtAuthenticationFilter' defined in class path resource [org/airsonic/player/security/GlobalSecurityConfig$ExtSecurityConfiguration.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.airsonic.player.security.JWTRequestParameterProcessingFilter]: Factory method 'jwtAuthFilter' threw exception; nested exception is java.lang.IllegalStateException: Cannot apply org.springframework.security.config.annotation.authentication.configurers.ldap.LdapAuthenticationProviderConfigurer@773c2214 to already built object,
	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:656) ~[spring-beans-5.2.5.RELEASE.jar!/:5.2.5.RELEASE],
	at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:484) ~[spring-beans-5.2.5.RELEASE.jar!/:5.2.5.RELEASE],
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1338) ~[spring-beans-5.2.5.RELEASE.jar!/:5.2.5.RELEASE],
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1177) ~[spring-beans-5.2.5.RELEASE.jar!/:5.2.5.RELEASE],
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:557) ~[spring-beans-5.2.5.RELEASE.jar!/:5.2.5.RELEASE],
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:517) ~[spring-beans-5.2.5.RELEASE.jar!/:5.2.5.RELEASE],
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:323) ~[spring-beans-5.2.5.RELEASE.jar!/:5.2.5.RELEASE],
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:222) ~[spring-beans-5.2.5.RELEASE.jar!/:5.2.5.RELEASE],
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:321) ~[spring-beans-5.2.5.RELEASE.jar!/:5.2.5.RELEASE],
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:207) ~[spring-beans-5.2.5.RELEASE.jar!/:5.2.5.RELEASE],
	at org.springframework.boot.web.servlet.ServletContextInitializerBeans.getOrderedBeansOfType(ServletContextInitializerBeans.java:211) ~[spring-boot-2.2.6.RELEASE.jar!/:2.2.6.RELEASE],
	at org.springframework.boot.web.servlet.ServletContextInitializerBeans.addAsRegistrationBean(ServletContextInitializerBeans.java:174) ~[spring-boot-2.2.6.RELEASE.jar!/:2.2.6.RELEASE],
	at org.springframework.boot.web.servlet.ServletContextInitializerBeans.addAsRegistrationBean(ServletContextInitializerBeans.java:169) ~[spring-boot-2.2.6.RELEASE.jar!/:2.2.6.RELEASE],
	at org.springframework.boot.web.servlet.ServletContextInitializerBeans.addAdaptableBeans(ServletContextInitializerBeans.java:154) ~[spring-boot-2.2.6.RELEASE.jar!/:2.2.6.RELEASE],
	at org.springframework.boot.web.servlet.ServletContextInitializerBeans.<init>(ServletContextInitializerBeans.java:86) ~[spring-boot-2.2.6.RELEASE.jar!/:2.2.6.RELEASE],
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.getServletContextInitializerBeans(ServletWebServerApplicationContext.java:253) ~[spring-boot-2.2.6.RELEASE.jar!/:2.2.6.RELEASE],
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.selfInitialize(ServletWebServerApplicationContext.java:227) ~[spring-boot-2.2.6.RELEASE.jar!/:2.2.6.RELEASE],
	at org.springframework.boot.web.embedded.tomcat.TomcatStarter.onStartup(TomcatStarter.java:53) ~[spring-boot-2.2.6.RELEASE.jar!/:2.2.6.RELEASE],
	at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5140) ~[tomcat-embed-core-9.0.33.jar!/:9.0.33],
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183) ~[tomcat-embed-core-9.0.33.jar!/:9.0.33],
	at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1384) ~[tomcat-embed-core-9.0.33.jar!/:9.0.33],
	at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1374) ~[tomcat-embed-core-9.0.33.jar!/:9.0.33],
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source) ~[na:na],
	at org.apache.tomcat.util.threads.InlineExecutorService.execute(InlineExecutorService.java:75) ~[tomcat-embed-core-9.0.33.jar!/:9.0.33],
	at java.base/java.util.concurrent.AbstractExecutorService.submit(Unknown Source) ~[na:na],
	at org.apache.catalina.core.ContainerBase.startInternal(ContainerBase.java:909) ~[tomcat-embed-core-9.0.33.jar!/:9.0.33],
	at org.apache.catalina.core.StandardHost.startInternal(StandardHost.java:841) ~[tomcat-embed-core-9.0.33.jar!/:9.0.33],
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183) ~[tomcat-embed-core-9.0.33.jar!/:9.0.33],
	at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1384) ~[tomcat-embed-core-9.0.33.jar!/:9.0.33],
	at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1374) ~[tomcat-embed-core-9.0.33.jar!/:9.0.33],
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source) ~[na:na],
	at org.apache.tomcat.util.threads.InlineExecutorService.execute(InlineExecutorService.java:75) ~[tomcat-embed-core-9.0.33.jar!/:9.0.33],
	at java.base/java.util.concurrent.AbstractExecutorService.submit(Unknown Source) ~[na:na],
	at org.apache.catalina.core.ContainerBase.startInternal(ContainerBase.java:909) ~[tomcat-embed-core-9.0.33.jar!/:9.0.33],
	at org.apache.catalina.core.StandardEngine.startInternal(StandardEngine.java:262) ~[tomcat-embed-core-9.0.33.jar!/:9.0.33],
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183) ~[tomcat-embed-core-9.0.33.jar!/:9.0.33],
	at org.apache.catalina.core.StandardService.startInternal(StandardService.java:421) ~[tomcat-embed-core-9.0.33.jar!/:9.0.33],
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183) ~[tomcat-embed-core-9.0.33.jar!/:9.0.33],
	at org.apache.catalina.core.StandardServer.startInternal(StandardServer.java:930) ~[tomcat-embed-core-9.0.33.jar!/:9.0.33],
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183) ~[tomcat-embed-core-9.0.33.jar!/:9.0.33],
	at org.apache.catalina.startup.Tomcat.start(Tomcat.java:468) ~[tomcat-embed-core-9.0.33.jar!/:9.0.33],
	at org.springframework.boot.web.embedded.tomcat.TomcatWebServer.initialize(TomcatWebServer.java:107) ~[spring-boot-2.2.6.RELEASE.jar!/:2.2.6.RELEASE],
	... 20 common frames omitted,
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.airsonic.player.security.JWTRequestParameterProcessingFilter]: Factory method 'jwtAuthFilter' threw exception; nested exception is java.lang.IllegalStateException: Cannot apply org.springframework.security.config.annotation.authentication.configurers.ldap.LdapAuthenticationProviderConfigurer@773c2214 to already built object,
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:185) ~[spring-beans-5.2.5.RELEASE.jar!/:5.2.5.RELEASE],
	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:651) ~[spring-beans-5.2.5.RELEASE.jar!/:5.2.5.RELEASE],
	... 61 common frames omitted,
Caused by: java.lang.IllegalStateException: Cannot apply org.springframework.security.config.annotation.authentication.configurers.ldap.LdapAuthenticationProviderConfigurer@773c2214 to already built object,
	at org.springframework.security.config.annotation.AbstractConfiguredSecurityBuilder.add(AbstractConfiguredSecurityBuilder.java:195) ~[spring-security-config-5.2.2.RELEASE.jar!/:5.2.2.RELEASE],
	at org.springframework.security.config.annotation.AbstractConfiguredSecurityBuilder.apply(AbstractConfiguredSecurityBuilder.java:133) ~[spring-security-config-5.2.2.RELEASE.jar!/:5.2.2.RELEASE],
	at org.springframework.security.config.annotation.authentication.builders.AuthenticationManagerBuilder.ldapAuthentication(AuthenticationManagerBuilder.java:206) ~[spring-security-config-5.2.2.RELEASE.jar!/:5.2.2.RELEASE],
	at org.airsonic.player.security.GlobalSecurityConfig.configure(GlobalSecurityConfig.java:178) ~[classes!/:11.0.0-SNAPSHOT],
	at org.springframework.security.config.annotation.authentication.configuration.GlobalAuthenticationConfigurerAdapter.configure(GlobalAuthenticationConfigurerAdapter.java:33) ~[spring-security-config-5.2.2.RELEASE.jar!/:5.2.2.RELEASE],
	at org.airsonic.player.security.GlobalSecurityConfig$$EnhancerBySpringCGLIB$$ea632942.configure(<generated>) ~[classes!/:11.0.0-SNAPSHOT],
	at org.springframework.security.config.annotation.AbstractConfiguredSecurityBuilder.configure(AbstractConfiguredSecurityBuilder.java:383) ~[spring-security-config-5.2.2.RELEASE.jar!/:5.2.2.RELEASE],
	at org.springframework.security.config.annotation.AbstractConfiguredSecurityBuilder.doBuild(AbstractConfiguredSecurityBuilder.java:329) ~[spring-security-config-5.2.2.RELEASE.jar!/:5.2.2.RELEASE],
	at org.springframework.security.config.annotation.AbstractSecurityBuilder.build(AbstractSecurityBuilder.java:41) ~[spring-security-config-5.2.2.RELEASE.jar!/:5.2.2.RELEASE],
	at org.springframework.security.config.annotation.authentication.configuration.AuthenticationConfiguration.getAuthenticationManager(AuthenticationConfiguration.java:118) ~[spring-security-config-5.2.2.RELEASE.jar!/:5.2.2.RELEASE],
	at org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter.authenticationManager(WebSecurityConfigurerAdapter.java:269) ~[spring-security-config-5.2.2.RELEASE.jar!/:5.2.2.RELEASE],
	at org.airsonic.player.security.GlobalSecurityConfig$ExtSecurityConfiguration.jwtAuthFilter(GlobalSecurityConfig.java:208) ~[classes!/:11.0.0-SNAPSHOT],
	at org.airsonic.player.security.GlobalSecurityConfig$ExtSecurityConfiguration$$EnhancerBySpringCGLIB$$a96e334b.CGLIB$jwtAuthFilter$1(<generated>) ~[classes!/:11.0.0-SNAPSHOT],
	at org.airsonic.player.security.GlobalSecurityConfig$ExtSecurityConfiguration$$EnhancerBySpringCGLIB$$a96e334b$$FastClassBySpringCGLIB$$b4e7e43d.invoke(<generated>) ~[classes!/:11.0.0-SNAPSHOT],
	at org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:244) ~[spring-core-5.2.5.RELEASE.jar!/:5.2.5.RELEASE],
	at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:331) ~[spring-context-5.2.5.RELEASE.jar!/:5.2.5.RELEASE],
	at org.airsonic.player.security.GlobalSecurityConfig$ExtSecurityConfiguration$$EnhancerBySpringCGLIB$$a96e334b.jwtAuthFilter(<generated>) ~[classes!/:11.0.0-SNAPSHOT],
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na],
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source) ~[na:na],
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source) ~[na:na],
	at java.base/java.lang.reflect.Method.invoke(Unknown Source) ~[na:na],
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:154) ~[spring-beans-5.2.5.RELEASE.jar!/:5.2.5.RELEASE],
	... 62 common frames omitted,
,
```

--- a/airsonic-main/src/main/java/org/airsonic/player/controller/ExternalPlayerController.java
+++ b/airsonic-main/src/main/java/org/airsonic/player/controller/ExternalPlayerController.java
@@ -104,7 +104,7 @@ public class ExternalPlayerController {
         Instant expires = null;
         Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
         if (authentication instanceof JWTAuthenticationToken) {
-            DecodedJWT token = jwtSecurityService.verify((String) authentication.getCredentials());
+            DecodedJWT token = (DecodedJWT) authentication.getDetails();
             expires = Optional.ofNullable(token).map(x -> x.getExpiresAt()).map(x -> x.toInstant()).orElse(null);
         }
         Instant finalExpires = expires;

--- a/airsonic-main/src/main/java/org/airsonic/player/security/GlobalSecurityConfig.java
+++ b/airsonic-main/src/main/java/org/airsonic/player/security/GlobalSecurityConfig.java
@@ -19,7 +19,6 @@ import org.springframework.core.annotation.Order;
 import org.springframework.security.authentication.AbstractAuthenticationToken;
 import org.springframework.security.authentication.event.AbstractAuthenticationFailureEvent;
 import org.springframework.security.config.annotation.authentication.builders.AuthenticationManagerBuilder;
-import org.springframework.security.config.annotation.authentication.configuration.GlobalAuthenticationConfigurerAdapter;
 import org.springframework.security.config.annotation.method.configuration.EnableGlobalMethodSecurity;
 import org.springframework.security.config.annotation.web.builders.HttpSecurity;
 import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
@@ -46,7 +45,7 @@ import static org.airsonic.player.security.MultipleCredsMatchingAuthenticationPr
 @Configuration
 @Order(SecurityProperties.BASIC_AUTH_ORDER - 2)
 @EnableGlobalMethodSecurity(securedEnabled = true, prePostEnabled = true)
-public class GlobalSecurityConfig extends GlobalAuthenticationConfigurerAdapter {
+public class GlobalSecurityConfig {
 
     private static final Logger LOG = LoggerFactory.getLogger(GlobalSecurityConfig.class);
 
@@ -172,8 +171,8 @@ public class GlobalSecurityConfig extends GlobalAuthenticationConfigurerAdapter 
         }
     }
 
-    @Override
-    public void configure(AuthenticationManagerBuilder auth) throws Exception {
+    @Autowired
+    public void configureGlobal(AuthenticationManagerBuilder auth) throws Exception {
         if (settingsService.isLdapEnabled()) {
             auth.ldapAuthentication()
                     .contextSource()

--- a/airsonic-main/src/main/java/org/airsonic/player/security/JWTAuthenticationProvider.java
+++ b/airsonic-main/src/main/java/org/airsonic/player/security/JWTAuthenticationProvider.java
@@ -17,6 +17,7 @@ import org.springframework.security.core.authority.SimpleGrantedAuthority;
 import org.springframework.web.util.UriComponents;
 import org.springframework.web.util.UriComponentsBuilder;
 
+import java.time.Instant;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Objects;
@@ -40,6 +41,11 @@ public class JWTAuthenticationProvider implements AuthenticationProvider {
         }
         String rawToken = (String) auth.getCredentials();
         DecodedJWT token = JWTSecurityService.verify(jwtKey, rawToken);
+
+        if (!token.getExpiresAt().toInstant().isAfter(Instant.now())) {
+            throw new InsufficientAuthenticationException("Credentials have expired");
+        }
+
         Claim path = token.getClaim(JWTSecurityService.CLAIM_PATH);
 
         // TODO:AD This is super unfortunate, but not sure there is a better way when using JSP
@@ -53,7 +59,7 @@ public class JWTAuthenticationProvider implements AuthenticationProvider {
         List<GrantedAuthority> authorities = new ArrayList<>();
         authorities.add(new SimpleGrantedAuthority("IS_AUTHENTICATED_FULLY"));
         authorities.add(new SimpleGrantedAuthority("ROLE_TEMP"));
-        return new JWTAuthenticationToken(token.getSubject(), rawToken, authentication.getRequestedPath(), authorities);
+        return new JWTAuthenticationToken(token.getSubject(), rawToken, authentication.getRequestedPath(), authorities, token);
     }
 
     private static boolean roughlyEqual(String expectedRaw, String requestedPathRaw) {

--- a/airsonic-main/src/main/java/org/airsonic/player/security/JWTAuthenticationToken.java
+++ b/airsonic-main/src/main/java/org/airsonic/player/security/JWTAuthenticationToken.java
@@ -19,6 +19,12 @@ public class JWTAuthenticationToken extends UsernamePasswordAuthenticationToken 
         this.requestedPath = requestedPath;
     }
 
+    public JWTAuthenticationToken(Object principal, Object credentials, String requestedPath,
+            Collection<? extends GrantedAuthority> authorities, Object details) {
+        this(principal, credentials, requestedPath, authorities);
+        this.setDetails(details);
+    }
+
     public String getRequestedPath() {
         return requestedPath;
     }

--- a/airsonic-main/src/test/java/org/airsonic/player/dao/UserDaoTestCase.java
+++ b/airsonic-main/src/test/java/org/airsonic/player/dao/UserDaoTestCase.java
@@ -13,8 +13,11 @@ import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.dao.DataIntegrityViolationException;
 import org.springframework.transaction.annotation.Propagation;
 import org.springframework.transaction.annotation.Transactional;
+import org.springframework.util.comparator.NullSafeComparator;
 
 import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.Comparator;
 import java.util.Locale;
 import java.util.Set;
 
@@ -173,7 +176,18 @@ public class UserDaoTestCase extends DaoTestCaseBean2 {
         UserSettings settings = new UserSettings("sindre");
         userDao.updateUserSettings(settings);
         UserSettings userSettings = userDao.getUserSettings("sindre");
-        assertThat(userSettings).usingRecursiveComparison().isEqualTo(settings);
+        assertThat(userSettings).usingComparatorForType(new NullSafeComparator<Instant>(new Comparator<Instant>() {
+            // use a custom comparator to account for micro second differences
+            // (Mysql only stores floats in json to a certain value)
+            @Override
+            public int compare(Instant o1, Instant o2) {
+                if (o1.equals(o2) || Math.abs(ChronoUnit.MICROS.between(o1, o2)) <= 2) {
+                    return 0;
+                }
+                return o1.compareTo(o2);
+            }
+        }, true), Instant.class)
+                .usingRecursiveComparison().isEqualTo(settings);
 
         settings = new UserSettings("sindre");
         settings.setLocale(Locale.SIMPLIFIED_CHINESE);

